### PR TITLE
Support init-empty-region in tikv-ctl

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/server/debug.rs [hunk 1]
```diff
@@ -326,6 +328,141 @@ impl Debugger {
...
+        let key = keys::region_state_key(region_id);
+        if box_try!(kv.get_msg_cf::<RegionLocalState>(CF_RAFT, &key)).is_some() {
+            return Err(Error::Other(
+                "Store already has the RegionLocalState".into(),
+            ));
+        }
+        box_try!(kv_wb.put_msg_cf(kv_handle, &key, &region_state));
+
+        // RaftApplyState.
+        let apply_state = box_try!(init_apply_state(kv, region_state.get_region()));
```
Comment (BusyJay) [inline src/server/debug.rs]: Use `write_initial_apply_state` instead.

File: src/server/debug.rs [hunk 2]
```diff
@@ -326,6 +328,141 @@ impl Debugger {
...
+            return Err(Error::Other("Store already has the RaftApplyState".into()));
+        }
+        box_try!(kv_wb.put_msg_cf(kv_handle, &key, &apply_state));
+
+        // RaftLocalState.
+        let key = keys::raft_state_key(region_id);
+        if box_try!(raft.get_msg::<RaftLocalState>(&key)).is_some() {
+            return Err(Error::Other("Store already has the RaftLocalState".into()));
+        }
+        box_try!(init_raft_state(raft, region_state.get_region()));
```
Comment (BusyJay) [inline src/server/debug.rs]: Use `write_initial_raft_state` instead.

File: src/server/debug.rs [hunk 3]
```diff
@@ -326,6 +328,141 @@ impl Debugger {
...
+        // RegionLocalState.
+        let mut region_state = RegionLocalState::new();
+        region_state.set_state(PeerState::Normal);
+        {
+            let region = region_state.mut_region();
+            region.set_id(region_id);
+            region.set_start_key(start_key);
+            region.set_end_key(end_key);
+
+            region.mut_region_epoch().set_conf_ver(1);
```
Comment (BusyJay) [inline src/server/debug.rs]: Use `INIT_EPOCH_VER` instead.

File: src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -990,6 +1050,37 @@ fn main() {
...
+                        .short("r")
+                        .takes_value(true)
+                        .help("the region id"),
+                ),
+        )
+        .subcommand(
+            SubCommand::with_name("init-empty-region")
+                .about("init an empty region on the store")
+                .arg(
+                    Arg::with_name("pd")
```
Comment (huachaohuang) [inline src/bin/tikv-ctl.rs]: Is it better to make the PD endpoints a first level argument, like the host and certification arguments?

Comment (overvenus) [reply src/bin/tikv-ctl.rs]: Most APIs do not require PD.

File: src/server/debug.rs [hunk 5]
```diff
@@ -386,6 +388,101 @@ impl Debugger {
...
+    pub fn remove_region(&self, region_id: u64) -> Result<()> {
+        let kv = self.engines.kv_engine.as_ref();
+        let raft = self.engines.raft_engine.as_ref();
+        let kv_wb = WriteBatch::new();
+        let raft_wb = WriteBatch::new();
+
+        let key = keys::raft_state_key(region_id);
+        let raft_state = match box_try!(raft.get_msg::<RaftLocalState>(&key)) {
+            Some(s) => s,
+            None => return Err(Error::Other("No RaftLocalState".into())),
```
Comment (BusyJay) [inline src/server/debug.rs]: The error is not friendly.

File: src/server/debug.rs [hunk 6]
```diff
@@ -386,6 +388,101 @@ impl Debugger {
...
+                    return Ok(true);
+                }
+
+                let mut region_state = RegionLocalState::new();
+                box_try!(region_state.merge_from_bytes(value));
+                if region_state.get_state() == PeerState::Tombstone {
+                    return Ok(true);
+                }
+                let exists_region = region_state.get_region();
+                if exists_region.get_start_key() >= region.get_end_key()
```
Comment (BusyJay) [inline src/server/debug.rs]: This is wrong.

File: src/bin/tikv-ctl.rs (hunk 4 above)
Comment (hicqu) [reply src/bin/tikv-ctl.rs]: So I prefer keep it as a subcommand argument.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/server/debug.rs [hunk 7]
```diff
@@ -696,6 +808,48 @@ mod tests {
             .unwrap()
     }
 
+    #[test]
+    fn test_region_overlay() {
+        let region_with = |start: &[u8], end: &[u8]| -> Region {
```
Comment (BusyJay) [inline src/server/debug.rs]: s/region_with/new_region/

File: src/server/debug.rs [hunk 8]
```diff
@@ -429,6 +526,21 @@ impl Debugger {
     }
 }
 
+fn region_overlay(r1: &Region, r2: &Region) -> bool {
```
Comment (BusyJay) [inline src/server/debug.rs]: s/region_overlay/region_overlap/

File: src/server/debug.rs [hunk 9]
```diff
@@ -696,6 +808,48 @@ mod tests {
...
+        // For normal case.
+        assert!(region_overlay(
+            &region_with(b"a", b"z"),
+            &region_with(b"b", b"y")
+        ));
+        assert!(region_overlay(
+            &region_with(b"a", b"n"),
+            &region_with(b"m", b"z")
+        ));
+        assert!(region_overlay(
```
Comment (BusyJay) [inline src/server/debug.rs]: Why?

**Comment by hicqu** [comment]
@BusyJay PTAL again thanks.

File: src/bin/tikv-ctl.rs [hunk 10]
```diff
@@ -1021,6 +1082,37 @@ fn main() {
                     ),
                 ),
         )
+        .subcommand(
+            SubCommand::with_name("remove-region")
+                .about("remove a region")
+                .arg(
+                    Arg::with_name("region")
+                        .short("r")
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: Is it required?

File: src/bin/tikv-ctl.rs [hunk 11]
```diff
@@ -1021,6 +1082,37 @@ fn main() {
...
+                        .takes_value(true)
+                        .help("the region id"),
+                ),
+        )
+        .subcommand(
+            SubCommand::with_name("init-empty-region")
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: May not be a good name. The command will re-initialize the region with a new region id and peers.

File: src/bin/tikv-ctl.rs [hunk 12]
```diff
@@ -1021,6 +1082,37 @@ fn main() {
...
+                        .takes_value(true)
+                        .multiple(true)
+                        .use_delimiter(true)
+                        .require_delimiter(true)
+                        .value_delimiter(",")
+                        .help("PD endpoints"),
+                )
+                .arg(
+                    Arg::with_name("region")
+                        .short("r")
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: Is it required?

File: src/server/debug.rs [hunk 13]
```diff
@@ -386,6 +388,101 @@ impl Debugger {
...
+            raft,
+            &kv_wb,
+            &raft_wb,
+            region_id,
+            &raft_state
+        ));
+
+        let mut write_opts = WriteOptions::new();
+        write_opts.set_sync(true);
+        box_try!(kv.write_opt(kv_wb, &write_opts));
```
Comment (BusyJay) [inline src/server/debug.rs]: Tombstone state is not set.

File: src/server/debug.rs [hunk 14]
```diff
@@ -386,6 +388,101 @@ impl Debugger {
...
+        let raft = self.engines.raft_engine.as_ref();
+        let kv_wb = WriteBatch::new();
+        let raft_wb = WriteBatch::new();
+
+        let key = keys::raft_state_key(region_id);
+        let raft_state = match box_try!(raft.get_msg::<RaftLocalState>(&key)) {
+            Some(s) => s,
+            None => return Err(Error::Other("No RaftLocalState in RAFT rocksdb".into())),
+        };
+        box_try!(clear_meta(
```
Comment (BusyJay) [inline src/server/debug.rs]: Why not just write a tombstone state? They will be cleared when tikv is started.

File: src/server/debug.rs [hunk 15]
```diff
@@ -386,6 +388,101 @@ impl Debugger {
...
+                if region_state.get_state() == PeerState::Tombstone {
+                    return Ok(true);
+                }
+                let exists_region = region_state.get_region();
+
+                if !region_overlap(exists_region, &region) {
+                    return Ok(true);
+                }
+
+                Err(RaftstoreError::Other("region overlay".into()))
```
Comment (BusyJay) [inline src/server/debug.rs]: s/overlay/overlap/

File: src/server/debug.rs (hunk 14 above)
Comment (hicqu) [reply src/server/debug.rs]: In the original design `remove-region` is used for testing `init-empty-region`. If we just set it tombstone, we need to restart tikv twice for it.

File: src/bin/tikv-ctl.rs [hunk 16]
```diff
@@ -411,6 +411,12 @@ trait DebugExecutor {
     /// Recover the cluster when given `store_ids` are failed.
     fn remove_fail_stores(&self, store_ids: Vec<u64>);
 
+    /// Remove region.
```
Comment (Hoverbear) [inline src/bin/tikv-ctl.rs]: This can be removed since it is just a repeat of the function name.

File: src/bin/tikv-ctl.rs [hunk 17]
```diff
@@ -411,6 +411,12 @@ trait DebugExecutor {
...
+    fn remove_region(&self, region_id: u64);
+
+    /// Init a new empty region on the store with given infos.
+    fn init_empty_region(&self, Arc<SecurityManager>, &PdConfig, u64);
```
Comment (Hoverbear) [inline src/bin/tikv-ctl.rs]: It's not clear what the parameters mean here. Could we name them like in the functions above?

File: src/server/debug.rs [hunk 18]
```diff
@@ -700,6 +804,48 @@ mod tests {
             .unwrap()
     }
 
+    #[test]
+    fn test_region_overlap() {
```
Comment (Hoverbear) [inline src/server/debug.rs]: Really nice, concise test. 👍

File: src/server/debug.rs (hunk 13 above)
Comment (hicqu) [reply src/server/debug.rs]: Tombstone it is not my purpose. I think we need more comments for it.

Comment (hicqu) [reply src/server/debug.rs]: I have commented `fn remove-region`. Maybe it's better now.

**Comment by hicqu** [comment]
ping @Hoverbear PTAL again, thanks.

**Comment by hicqu** [comment]
ping @BusyJay

**Comment by hicqu** [comment]
ping @BusyJay

File: src/server/debug.rs [hunk 19]
```diff
@@ -431,6 +433,78 @@ impl Debugger {
...
+                }
+
+                let mut region_state = RegionLocalState::new();
+                box_try!(region_state.merge_from_bytes(value));
+                if region_state.get_state() == PeerState::Tombstone {
+                    return Ok(true);
+                }
+                let exists_region = region_state.get_region();
+
+                if !region_overlap(exists_region, &region) {
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/server/debug.rs]: Fixed. PTAL.

**Comment by hicqu** [comment]
ping @BusyJay

**Comment by hicqu** [comment]
friendly ping @BusyJay .

**Comment by BusyJay** [comment]
Please resolve the conflict and get it merged.

**Pull request merged**